zerver/views/auth.py

```diff
@@ -386,12 +382,16 @@ def start_social_login(request: HttpRequest, backend: str) -> HttpResponse:
 
     return oauth_redirect_to_root(request, backend_url, 'social', extra_url_params=extra_url_params)
 
-def start_social_signup(request: HttpRequest, backend: str) -> HttpResponse:
+def start_social_signup(request: HttpRequest, backend: str, extra_arg: Optional[str]=None
+                        ) -> HttpResponse:
     backend_url = reverse('social:begin', args=[backend])
     extra_url_params = {}  # type: Dict[str, str]
     if backend == "saml":
-        assert len(settings.SOCIAL_AUTH_SAML_ENABLED_IDPS) == 1
-        extra_url_params = {'idp': list(settings.SOCIAL_AUTH_SAML_ENABLED_IDPS.keys())[0]}
+        if not extra_arg or extra_arg not in settings.SOCIAL_AUTH_SAML_ENABLED_IDPS:
+            logging.info("Attempted to initiate SAML authentication with wrong idp argument: {}"
+                         .format(extra_arg))
+            return redirect_to_config_error("saml")
+        extra_url_params = {'idp': extra_arg}
```

Comment:
This can be a follow-up commit, but I think this block should match the similar block in the login section (i.e. include testing that the SAML settings are configured, etc.).  I think that means we should just make it a shared function that returns an `Optional[HttpResponse]` and then have both this and `start_social_login` call that function, check the return value, and if it's not None, return it.

Reply:
Changed in new commit `` saml: Sanity-check configuration in both login and signup codepaths.``